sdk/src/azure/iot/az_iot_hub_client.c

```diff
@@ -61,27 +64,38 @@ AZ_NODISCARD az_result az_iot_hub_client_get_user_name(
   const az_span* const module_id = &(client->_internal.options.module_id);
   const az_span* const user_agent = &(client->_internal.options.user_agent);
   const az_span* const model_id = &(client->_internal.options.model_id);
+  const az_span* const method_twin_ct = &(client->_internal.options.method_twin_content_type);
 
   az_span mqtt_user_name_span
       = az_span_create((uint8_t*)mqtt_user_name, (int32_t)mqtt_user_name_size);
 
   int32_t required_length = az_span_size(client->_internal.iot_hub_hostname)
-      + az_span_size(client->_internal.device_id) + (int32_t)sizeof(hub_client_forward_slash)
+      + (int32_t)sizeof(hub_client_forward_slash) + az_span_size(client->_internal.device_id)
       + az_span_size(hub_service_api_version);
+
   if (az_span_size(*module_id) > 0)
   {
-    required_length += az_span_size(*module_id) + (int32_t)sizeof(hub_client_forward_slash);
+    required_length += (int32_t)sizeof(hub_client_forward_slash) + az_span_size(*module_id);
   }
   if (az_span_size(*user_agent) > 0)
   {
-    required_length += az_span_size(*user_agent) + az_span_size(hub_client_param_separator_span);
+    required_length += az_span_size(hub_client_param_separator_span) + az_span_size(*user_agent);
   }
-  // Note we skip the length of the model id since we have to url encode it. Bound checking is done
-  // later.
   if (az_span_size(*model_id) > 0)
+  {
+    // Temporarily use mqtt_user_name_span. If not big enough for encoded_length, return error here.
+    int32_t encoded_length = 0;
+    _az_RETURN_IF_FAILED(_az_span_url_encode(mqtt_user_name_span, *model_id, &encoded_length));
```

Comment:
I am just thinking out loud here, but doesn't overriding `mqtt_user_name_span` while doing url encoding result in the initial data being trampled over that the user provided, since url encoding typically increases the size of the data?

Why is that OK to do?

Reply:
`mqtt_user_name_span` is created but not yet used, so it won't be trampling over anything.  I considered creating a different az_span of the same size just to get the encoded length, but thought that was redundant.  However, it could prevent against a future bug if someone didn't catch this nuance.

Reply:
> `mqtt_user_name_span` is created but not yet used, so it won't be trampling over anything. 

The underlying buffer that we are writing to (that you created the span out of) is provided by the user. Do they have data in there that we need?

Reply:
The problem is that if we url encode and check the required total length later (when model-id is added to the username), then we aren't getting an accurate url required length check since the content type has to go after the model id.

The first required length check is done sans url-encoded model id.
The username is constructed with url-encoded model id, and the content type is added after that.  
There is no accurate required length check done, unless we do it AGAIN.

Reply:
I have rewritten the logic to do pre-url encode size bound checking and building of username, then logic for the model id url encoding, and then a post-url encode size bound checking and building of username.

Reply:
A different option is to bound check size and build the username for each item as we go along.  
Or, build the complete username, and bound check at the end.

Reply:
@jspaith, interested in what you think about a best approach here.

Reply:
> Unless, its ok for model_id to go at the end.

Yes that should be fine since they are really just url parameters. Hub will parse them out wherever they are. I would put the model-id at the end and check the size of everything before that url encode section.

Reply:
After thinking about it, for now I'm going to keep it with the order specified in the docs for consistency.

Reply:
Will try this out.

Reply:
Fixed, and tests updated.